docs/notebooks/jax-101/01-jax-basics.md

```diff
@@ -0,0 +1,375 @@
+:id: dKeorwJfvpeI
+
+sum_of_squares_dx = jax.grad(sum_of_squares)
+
+x = jnp.asarray([1.0, 2.0, 3.0, 4.0])
+
+print(sum_of_squares(x))
+
+print(sum_of_squares_dx(x))
+```
+
++++ {"id": "VfBt5CYbyKUX"}
+
+You can think of `jax.grad` by analogy to the $\nabla$ operator from vector calculus. Given a function $f(x)$, $\nabla f$ represents the function that computes $f$'s gradient, i.e.
+
+$$
+(\nabla f)(x)_i = \frac{\partial f}{\partial x_i}(x).
+$$
+
+Analogously, `jax.grad(f)` is the function that computes the gradient, so `jax.grad(f)(x)` is the gradient of `f` at `x`.
+
+(Like $\nabla$, `jax.grad` will only work on functions with a scalar output -- it will raise an error otherwise.)
+
+This makes the JAX API quite different to other autodiff libraries like Tensorflow and PyTorch, where to compute the gradient we use the loss tensor itself (e.g. by calling `loss.backward()`). The JAX API works directly with functions, staying closer to the underlying math. Once you become accustomed to this way of doing things, it feels natural: your loss function in code really is a function of parameters and data, and you find its gradient just like you would in the math.
+
+This way of doing things makes it straightforward to control things like which variables to differentiate with respect to. By default, `jax.grad` will find the gradient with respect to the first argument. In the example below, the result of `sum_squared_error_dx` will be the gradient of `sum_squared_error` with respect to `x`.
+
+```{code-cell} ipython3
+:id: f3NfaVu4yrQE
+
+def sum_squared_error(x, y):
+  return jnp.sum((x-y)**2)
+
+sum_squared_error_dx = jax.grad(sum_squared_error)
+
+y = jnp.asarray([1.1, 2.1, 3.1, 4.1])
+
+print(sum_squared_error_dx(x, y))
+```
+
++++ {"id": "1tOztA5zpLWN"}
+
+To find the gradient with respect to a different argument (or several), you can set `argnums`:
+
+```{code-cell} ipython3
+:id: FQSczVQkqIPY
+
+jax.grad(sum_squared_error, argnums=(0, 1))(x, y)  # Find gradient wrt both x & y
+```
+
++++ {"id": "yQAMTnZSqo-t"}
+
+Does this mean that when doing machine learning, we need to write functions with gigantic argument lists, with an argument for each model parameter array? No. JAX comes equipped with machinery for bundling arrays together in data structures called 'pytrees', on which more in a [later guide](https://colab.research.google.com/github/google/jax/blob/master/docs/notebooks/jax-101/06-pytrees.ipynb). So, most often, use of `jax.grad` looks like this:
+
+```
+def loss_fn(params, data):
+  ...
+
+grads = jax.grad(loss_fn)(params, data_batch)
+```
+
++++ {"id": "oBowiovisT97"}
+
+where `params` is, for example, a nested dict of arrays, and the returned `grads` is another nested dict of arrays with the same structure.
+
++++ {"id": "LNjf9jUEsZZ8"}
+
+## Value and Grad
+
+Often, you need to find both the value and the gradient of a function, e.g. if you want to log the training loss. JAX has a handy sister transformation for efficiently doing that:
```

Comment:
Suggestion:

- Since it's a sister JAX transform, maybe we should mention it before "revealing" it in a code snippet:

```suggestion
Often, you need to find both the value and the gradient of a function, e.g. if you want to log the training loss. JAX has a handy sister transformation—`value_and_grad`—for efficiently doing that:
```

(or something along those lines)